### PR TITLE
[WIP][JIT] Vectorizing fusion compiler for CPU

### PR DESCRIPTION
**Pull request by jamesr66a**
Stacked on top of https://github.com/pytorch/pytorch/pull/18819

Brain dump to get some design discussions going. Just something I hacked together to see if we can get some wins. It gets a 7.8x win for batch norm on my devserver on a 168kB input tensor.

TODOs:
- Figure out the `-march=native` thing, where calling `g++` with that flag is currently unconditionally disabled. Maybe we can guard it behind an environment variable

cc @zdevito

### CODE REVIEW

**Comment by soumith** [comment]
> Can we package and distribute Vec256 s.t. the fusion compiler can use it?

Absolutely, yes. This must've been a cmake oversight

**Comment by soumith** [comment]
cc: @cpuhrsch @colesbury

**Comment by jamesr66a** [comment]
@soumith, sorry, I mean we need to figure out how to package up Vec256 so that it can be used when we call out to `g++` in the fusion compiler. Not if it's packaged up in PyTorch

File: torch/csrc/jit/fuser/cpu/resource_strings.h [hunk 1]
```diff
@@ -44,10 +95,24 @@ scalar_t rsqrtf(scalar_t x) {
 
 #define OMP_THRESHOLD 100000
 static void ${kernelName}_kernel(IndexType totalElements, ${formals}) {
-  #pragma omp parallel for if(totalElements > OMP_THRESHOLD)
-  for (IndexType linearIndex = 0;
+  IndexType linearIndex = 0;
+  if (/*vectorizable=*/${vectorizable}) {
```
Comment (zdevito) [inline torch/csrc/jit/fuser/cpu/resource_strings.h]: might have to make this a formatting-time distinction if  the vector instructions don't exist.

File: torch/csrc/jit/fuser/codegen.cpp [hunk 2]
```diff
@@ -397,13 +463,26 @@ std::string generateKernel(
...
     } else {
       env.s("access", format("s${formal}", env));
       env.s("lhs_type", variableType(input.first->type()));
+      if (vectorizable) {
+        env.s("vector_access", format("Vec256(s${formal})", env));
+        env.s("vector_lhs_type", "Vec256");
+      }
     }
     body << format("${lhs_type} ${node} = ${access};\n", env);
+    if (vectorizable) {
```
Comment (zdevito) [inline torch/csrc/jit/fuser/codegen.cpp]: probably don't need to guard this stuff as long as you are writing to vector-specific strings anyway.
